subsystem/apiclient/kubeapi.go

```diff
@@ -0,0 +1,288 @@
+		},
+		Data: secretData,
+	}
+	if err := kc.client.Create(ctx, secret); err != nil {
+		return nil, err
+	}
+
+	return secretRef, nil
+}
+
+func (kc *KubeAPIClient) verifyClusterStatus(ctx context.Context, key types.NamespacedName) error {
+	cluster := &adiiov1alpha1.Cluster{}
+	if err := kc.client.Get(ctx, key, cluster); err != nil {
+		return err
+	}
+	if cluster.Status.Error != "" {
+		return errors.Errorf("cluster status error: %s", cluster.Status.Error)
+	}
+	return nil
+}
+
+func (kc *KubeAPIClient) getClusterWithRetries(
+	ctx context.Context,
+	key types.NamespacedName,
+) (*models.Cluster, error) {
+
+	var cluster *models.Cluster
+	var getClusterErr error
+	for i := 0; i < 10; i++ {
+		if verifyStatusErr := kc.verifyClusterStatus(ctx, key); verifyStatusErr != nil {
+			return nil, verifyStatusErr
+		}
+		if cluster, getClusterErr = kc.getClusterByKey(key); getClusterErr == gorm.ErrRecordNotFound {
+			time.Sleep(time.Millisecond * 500)
+			continue
+		}
+		break
+	}
+	if getClusterErr != nil {
+		return nil, errors.Wrapf(getClusterErr, "failed to get cluster")
+	}
+	return cluster, nil
+}
+
+func (kc *KubeAPIClient) getClusterByKey(key types.NamespacedName) (*models.Cluster, error) {
+	c := &common.Cluster{}
+	if res := kc.db.Take(&c, "kube_key_name = ? and kube_key_namespace = ?",
+		key.Name, key.Namespace); res.Error != nil {
+		return nil, res.Error
+	}
+	return &c.Cluster, nil
+}
+
+func (kc *KubeAPIClient) DeregisterCluster(
+	ctx context.Context,
+	params *installer.DeregisterClusterParams,
+) error {
+
+	cluster, getErr := kc.getClusterByID(params.ClusterID)
+	if getErr != nil {
+		return errors.Wrapf(getErr, "failed getting cluster from db")
+	}
+	ref := &corev1.ObjectReference{
+		Name:      cluster.Name,
+		Namespace: kc.namespace,
+	}
+	if delErr := kc.client.Delete(ctx, ref); delErr != nil {
+		return errors.Wrapf(delErr, "failed deleting cluster")
+	}
+	return nil
+}
+
+func (kc *KubeAPIClient) getClusterByID(id strfmt.UUID) (*models.Cluster, error) {
+	c := &common.Cluster{}
+	if res := kc.db.Take(c, "id = ?", id.String()); res.Error != nil {
+		return nil, res.Error
+	} else if res.RowsAffected == 0 {
+		return nil, errors.New("cluster was not found in db")
+	}
+	return &c.Cluster, nil
+}
+
+func (kc *KubeAPIClient) UpdateCluster(
+	ctx context.Context,
+	params *installer.UpdateClusterParams,
+) (*models.Cluster, error) {
+
+	cParams := params.ClusterUpdateParams
+
+	clusterName := swag.StringValue(cParams.Name)
+
+	pullSecretRef, pullSecretErr := kc.deployPullSecret(
+		ctx,
+		clusterName,
+		swag.StringValue(cParams.PullSecret))
+	if pullSecretErr != nil {
+		return nil, errors.Wrapf(pullSecretErr, "failed to deploy pull secret")
+	}
+
+	metaName := uuid.New().String()
+	c := adiiov1alpha1.Cluster{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Cluster",
+			APIVersion: APIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      metaName,
+			Namespace: kc.namespace,
+		},
+		Spec: adiiov1alpha1.ClusterSpec{
```

Comment:
I think that some fields are missing here, for example `MachineNetworkCIDR` etc.